File: mahaam-api-cs/Src/Infra/Middleware.cs
using System.Diagnostics;
using System.Text;
using Mahaam.Infra.Monitoring;

namespace Mahaam.Infra;

public class AppMiddleware(RequestDelegate next)
{
	private readonly RequestDelegate _next = next;

	public async Task Invoke(HttpContext context)
	{
		var stopwatch = new Stopwatch();
		stopwatch.Start();
		Req.TrafficId = Guid.NewGuid();

		string? request = null;
		string? response = null;

		try
		{

			if (Config.LogReqEnabled)
			{
				// expensive operation, only used if needed
				context.Request.EnableBuffering();
			}
			AuthenticateReq(context);
			await _next(context);
		}
		catch (Exception e)
		{
			if (Config.LogReqEnabled)
			{
				context.Request.Body.Position = 0;
				request = await GetReqBody(context.Request);
			}
			response = await HandleException(context, e);
		}
		finally
		{
			var path = context.Request.Path.Value ?? "";
			var notTrafficPaths = path.StartsWith("/swagger") || path.Equals("/health") || path.StartsWith("/audit");
			if (!notTrafficPaths)
			{
				CreateTraffic(context, request, response, stopwatch);
			}
		}
	}

	private static void AuthenticateReq(HttpContext context)
	{
		var path = context.Request.Path.Value ?? "";
		var pathBase = context.Request.PathBase.Value;
		if (!"/mahaam-api".Equals(pathBase))
		{
			throw new NotFoundException("Invalid path base");
		}

		string? appStore = context.Request.Headers["x-app-store"];
		string? appVersion = context.Request.Headers["x-app-version"];
		if ((appStore == null || appVersion == null) && !path.StartsWith("/swagger"))
		{
			Log.Error($"Required headers not exists, appStore: {appStore}, appVersion: {appVersion}, path: {path}");
			throw new UnauthorizedException("Required headers not exists");
		}
		Req.AppStore = appStore!;
		Req.AppVersion = appVersion!;

		var bypassAuthPaths = new List<string> {
			"/swagger",
			"/health",
			"/audit",
			"/users/create"
		};

		if (!bypassAuthPaths.Exists(path.StartsWith))
		{
			(Guid userId, Guid deviceId, bool isLoggedIn) = Auth.ValidateAndExtractJwt(context);
			Req.UserId = userId;
			Req.DeviceId = deviceId;
			Req.IsLoggedIn = isLoggedIn;
		}
	}

	private static async Task<string> HandleException(HttpContext context, Exception e)
	{
		var response = Json.Serialize(e.Message);
		var code = Http.ServerError;


		if (e is AppException)
		{
			var appException = e as AppException;
			var key = appException!.Key;
			code = appException.HttpCode;
			if (!string.IsNullOrEmpty(key))
			{
				var res = new { key, error = e.Message };
				response = Json.Serialize(res);
			}
		}

		Log.Error(e.ToString());
		context.Response.StatusCode = code;
		context.Response.ContentType = Http.json;
		await context.Response.WriteAsync(response);
		return response;
	}

	private static void CreateTraffic(HttpContext context, string? request, string? response, Stopwatch stopwatch)
	{
		var method = context.Request.Method;
		var path = $"{context.Request.Path.Value}{context.Request.QueryString.ToString()}";
		var code = context.Response.StatusCode;
		var elapsed = stopwatch.ElapsedMilliseconds;
		var headers = new TrafficHeaders
		{
			UserId = Req.UserId,
			DeviceId = Req.DeviceId,
			AppStore = Req.AppStore,
			AppVersion = Req.AppVersion,
		};
		var isSuccessResponse = code < 400;
		var isUserPath = path.StartsWith("/user");
		if (string.IsNullOrEmpty(response) || isUserPath)
		{
			response = null;
		}

		if (isSuccessResponse)
		{
			request = null;
			response = null;
		}

		stopwatch.Stop();

		var traffic = new Traffic
		{
			Id = Req.TrafficId,
			Method = method,
			Path = path,
			Code = code,
			Elapsed = elapsed,
			Headers = Json.Serialize(headers),
			Request = request,
			Response = string.IsNullOrEmpty(response) ? null : response,
			HealthId = Cache.HealthId
		};
		App.TrafficRepo.Create(traffic);
	}

	private static async Task<string?> GetReqBody(HttpRequest request)
	{
		using var reader = new StreamReader(request.Body, Encoding.UTF8, true, 1024, true);
		var body = await reader.ReadToEndAsync();
		return string.IsNullOrEmpty(body) ? null : body;
	}
}

File: mahaam-api-ts/src/infra/middleware.ts
import {
  Injectable,
  NestMiddleware,
  HttpStatus,
  Inject,
  NestInterceptor,
  CallHandler,
  ExecutionContext,
} from '@nestjs/common';
import { Request, Response, NextFunction } from 'express';
import { randomUUID } from 'crypto';
import { Auth } from './auth';
import { UnauthorizedError, NotFoundError, AppError } from './errors';
import { Req } from './req';
import Log from './log';
import { Cache } from './cache';
import { Traffic } from './monitor/monitor.model';
import { TrafficRepo } from './monitor/traffic.repo';

import config from './config';
import { UsersRepo } from 'src/feat/users/users.repo';
import { DeviceRepo } from 'src/feat/users/device.repo';
import { Observable, throwError } from 'rxjs';
import { catchError, finalize, tap } from 'rxjs/operators';

@Injectable()
export class AppMiddleware implements NestMiddleware {
  constructor(
    @Inject('UsersRepo') private readonly usersRepo: UsersRepo,
    @Inject('DeviceRepo') private readonly deviceRepo: DeviceRepo,
  ) {}

  async use(req: Request, res: Response, next: NextFunction): Promise<void> {
    return Req.run(async () => {
      Req.startTime = Date.now();
      Req.trafficId = randomUUID();
      await this.authenticateReq(req);
      next();
    });
  }

  private async authenticateReq(req: Request): Promise<void> {
    let path = '';
    if (config.baseUrl && req.baseUrl.slice(1).startsWith(config.baseUrl)) {
      path = req.baseUrl.slice(config.baseUrl.length + 1);
    }

    if (config.baseUrl && !req.baseUrl.slice(1).startsWith(config.baseUrl)) {
      throw new NotFoundError('Invalid path base');
    }

    const appStore = req.headers['x-app-store'] as string;
    const appVersion = req.headers['x-app-version'] as string;

    if ((!appStore || !appVersion) && !path.startsWith('/swagger')) {
      throw new UnauthorizedError('Required headers not exists');
    }

    Req.appStore = appStore;
    Req.appVersion = appVersion;

    const bypassAuthPaths = ['/swagger', '/health', '/audit', '/users/create'];

    if (!bypassAuthPaths.some((bypassPath) => path.startsWith(bypassPath))) {
      const { userId, deviceId, isLoggedIn } = await Auth.validateAndExtractJwt(req, this.usersRepo, this.deviceRepo);
      Req.userId = userId;
      Req.deviceId = deviceId;
      Req.isLoggedIn = isLoggedIn;
    }
  }
}

@Injectable()
export class TrafficInterceptor implements NestInterceptor {
  constructor(@Inject('TrafficRepo') private readonly trafficRepo: TrafficRepo) {}

  intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
    let request: Request;
    let response: Response;
    let responseBody: string;

    return next.handle().pipe(
      catchError((err: unknown) => {
        Log.error(err instanceof Error ? err.toString() : String(err));

        const ctx = context.switchToHttp();
        response = ctx.getResponse<Response>();

        responseBody = JSON.stringify(err instanceof Error ? err.message : String(err));
        let code = HttpStatus.INTERNAL_SERVER_ERROR;

        if (err instanceof AppError) {
          const appException = err as AppError;
          const key = appException.key;
          code = appException.httpCode;
          if (key) {
            const responseObj = { key, error: err.message };
            responseBody = JSON.stringify(responseObj);
          }
        }
        response.status(code).contentType('application/json');
        response.send(responseBody);

        return throwError(() => err);
      }),
      finalize(() => {
        const ctx = context.switchToHttp();
        request = ctx.getRequest<Request>();
        response = ctx.getResponse<Response>();
        this.createTraffic(request, response, responseBody);
      }),
    );
  }

  private createTraffic(req: Request, res: Response, responseBody: string | null): void {
    const path = req.path;
    const notTrafficPaths = path.startsWith('/swagger') || path === '/health' || path.startsWith('/audit');
    if (notTrafficPaths) return;

    const elapsed = Date.now() - Req.startTime;

    const headers = {
      userId: Req.userId,
      deviceId: Req.deviceId,
      appStore: Req.appStore,
      appVersion: Req.appVersion,
    };

    const isSuccessResponse = res.statusCode < 400;
    const isUserPath = path.startsWith('/user');
    let requestBody: string | null = null;
    if (isUserPath) {
      responseBody = null;
    }

    if (!isSuccessResponse && config.logReqEnabled) {
      requestBody = this.getReqBody(req);
    }

    const traffic: Traffic = {
      id: Req.trafficId || randomUUID(),
      method: req.method,
      path,
      code: res.statusCode,
      elapsed,
      headers: JSON.stringify(headers),
      request: requestBody,
      response: responseBody,
      healthId: Cache.getHealthId(),
    };

    this.trafficRepo.create(traffic);
  }

  private getReqBody(request: Request): string | null {
    if (request.body) {
      const body = typeof request.body === 'string' ? request.body : JSON.stringify(request.body);
      return body || null;
    }
    return null;
  }
}
